### PR TITLE
Improve Jitify performance

### PR DESCRIPTION
**Pull request by leofang**
~~Blocked by my upstream PR NVIDIA/jitify#82~~, which allows reusing the already-constructed header map (in C++). Also, I eject all input code from this map to only keep headers. Together, this helps reduce the runtime for both the first shot compilation as well as a long-running program (such as running the test suite).

### CODE REVIEW

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
The upstream PR NVIDIA/jitify#82 is merged, and the submodule is updated accordingly. This is ready now @emcastillo.

File: cupy/cuda/compiler.py [hunk 1]
```diff
@@ -172,12 +176,16 @@ def _jitify_prep(source, options, cu_path):
...
             source, options, _jitify_header_source_map)
     except Exception as e:  # C++ could throw all kinds of errors
-        raise JitifyException(str(e))
+        cex = CompileException(str(e), source, cu_path, options, 'jitify')
+        dump = _get_bool_env_variable(
+            'CUPY_DUMP_CUDA_SOURCE_ON_ERROR', False)
+        if dump:
+            cex.dump(sys.stderr)
+        raise JitifyException(str(cex))
     assert name == cu_path
```
Comment (leofang) [inline cupy/cuda/compiler.py:188]: This line ensures that the upstream change (NVIDIA/jitify#82) is correctly behaving.

File: cupy/cuda/compiler.py [hunk 2]
```diff
@@ -172,12 +176,16 @@ def _jitify_prep(source, options, cu_path):
...
+        name, options, headers, include_names = jitify(
             source, options, _jitify_header_source_map)
     except Exception as e:  # C++ could throw all kinds of errors
-        raise JitifyException(str(e))
+        cex = CompileException(str(e), source, cu_path, options, 'jitify')
+        dump = _get_bool_env_variable(
+            'CUPY_DUMP_CUDA_SOURCE_ON_ERROR', False)
+        if dump:
+            cex.dump(sys.stderr)
+        raise JitifyException(str(cex))
```
Comment (leofang) [inline cupy/cuda/compiler.py:187]: This change allows the code (`source`) to be dumped if Jitify errors out, as sometimes it's not trivial to diagnose from the error log.

**Comment by leofang** [comment]
Jenkins, test this please

**Review by emcastillo — APPROVED** [review]

**Pull request merged**